library/std/src/sys/sync/rwlock/futex.rs

```diff
@@ -164,6 +185,27 @@ impl RwLock {
         }
     }
 
+    #[inline]
+    pub unsafe fn downgrade(&self) {
+        // Removes all the write bits and adds a single read bit.
+        let old_state = self.state.fetch_sub(WRITE_LOCKED - READ_LOCKED, Relaxed);
```

Comment:
Since `WRITE_LOCKED - READ_LOCKED` is used twice (once here once inverted a couple lines below), maybe it should get its own constant?

Reply:
Hmm that makes sense, but I'm not sure what we would call it, or if we would even place it next to the rest of the constants at the top of the file. Would we call it something like `const DOWNGRADE: u32 = WRITE_LOCKED - READ_LOCKED`? That wouldn't make a lot of sense since a downgrade should really be adding `READ_LOCKED - WRITE_LOCKED`, but I don't know if we should allow that overflow to happen, both in the constant itself and in the potential `fetch_add` addition?

Reply:
I came up with this:

```rust
const DOWNGRADE: u32 = READ_LOCKED.overflowing_sub(WRITE_LOCKED).0;
// <-- snip -->
    #[inline]
    pub unsafe fn downgrade(&self) {
        // Removes all the write bits and adds a single read bit.
        let old_state = self.state.fetch_add(DOWNGRADE, Relaxed);
        debug_assert!(
            is_write_locked(old_state),
            "RwLock must be write locked to call `downgrade`"
        );

        let state = old_state.overflowing_add(DOWNGRADE).0;
// <-- snip -->
```

I don't know if this is more or less clear, to be honest.

Reply:
Alternatively, I don't think the debug assertion makes that much sense, I think it makes the code harder to follow while not really checking much (it just checks the arithmetic operation right above). If we just removed that one then the `state = old_state - WRITE_LOCKED + READ_LOCKED` can be deleted and there's no real need for a constant

Reply:
Yep, that makes perfect sense. I'll remove that second debug assert